nbviewer/app.py

```diff
@@ -159,106 +172,137 @@ def init_tornado_application(self):
             git_data = {}
         else:
             git_data['msg'] = escape(git_data['msg'])
-    
-    
+
         if options.no_cache:
-            # force jinja to recompile template every time
+            # force Jinja2 to recompile template every time
             env.globals.update(cache_size=0)
-        env.globals.update(nrhead=nrhead, nrfoot=nrfoot, git_data=git_data,
-            jupyter_info=jupyter_info(), len=len,
-        )
-        AsyncHTTPClient.configure(HTTPClientClass)
-        client = AsyncHTTPClient()
-        client.cache = cache
-    
-        # load frontpage sections
-        with io.open(options.frontpage, 'r') as f:
-            frontpage_setup = json.load(f)
-        # check if the json has a 'sections' field, otherwise assume it is
-        # just a list of sessions, and provide the defaults for the other
-        # fields
-        if 'sections' not in frontpage_setup:
-            frontpage_setup = {'title': 'nbviewer',
-                               'subtitle':
-                               'A simple way to share Jupyter Notebooks',
-                               'show_input': True,
-                               'sections': frontpage_setup}
-    
-        # cache frontpage links for the maximum allowed time
-        max_cache_uris = {''}
-        for section in frontpage_setup['sections']:
-            for link in section['links']:
-                max_cache_uris.add('/' + link['target'])
-    
+        env.globals.update(nrhead=nrhead, nrfoot=nrfoot, git_data=git_data, jupyter_info=jupyter_info(), len=len)
```

Comment:
I'm wondering about the use of properties throughout the new code. 

In the original code, various values were computed and objects instantiated once-and-only-once in the top-to-bottom flow of `init_tornado_application`. Now many of the values have the potential to be computed more than once if their properties are accessed more than one time. For example, if we forget or a new developer doesn't know that`self.env` is a property, each time it's referenced the code will create a fresh instance of a `jinja2.Environment` and potentially configure it in a different manner. 

Maybe it doesn't matter if we build a new `env` each time. But I'm perhaps it will in the future. Or perhaps it does already matter for other properties (`pool`?) I suspect these properties acting like factory functions will become hard to maintain and understand over time.

Two thoughts on what we could do instead:

1. Have the properties cache their computed values on first access and only return the cached value there after.
2. Make the properties regular functions so that it's clear an explicit action is being taken each time one is called.

What do you think?

Reply:
My understanding was that properties are only computed once, so that they don't implement an observer pattern by default. So I thought this was logically equivalent to the original code. At the very least (I think) I know that for traitlets, the very similar `@default` decorator only computes the value once, and that one has to manually add an `@observe` decorator to implement an observer pattern. So maybe I should change these to traitlets?

I feel like I'm misunderstanding the concern.

Reply:
I guess I should have used setters instead of properties? https://www.python-course.eu/python3_properties.php

My rough understanding was that a property could safely be thought of as an attribute whose (initial and final for all time) value could be computed dynamically, using the function decorated by the decorator, and that therefore the property was effectively "read-only". But the more I look into this, the less this seems like this is the case. 

I know that traitlets attributes can be over-written outside of the class where they are defined, but they are at least "read-only" within the class where they are defined. (At least that has been my practical experience.) So maybe I should make these traitlets?

Reply:
Oh crud yeah I didn't understand how properties work at all -- the function is actually called every time the property is accessed, not just once: https://wiki.python.org/moin/ComputedAttributesUsingPropertyObjects

This behavior wasn't intentional on my part. I think option **1.** you listed above is the better one (if for no other reason than that was what I had originally intended anyway). Can that be implemented using traitlets and the `@default` decorator? 

I'm not sure how else that might be done. I thought the purpose of properties was for this sort of code organization, not to recompute the value every single time.

Reply:
Also if properties work that way, then it would seem like they automatically implement the observer pattern. But I thought I read elsewhere that properties on their own weren't sufficient to implement the observer pattern, which I think was part of what led to my mistaken belief about how they work. So basically now I'm confused. I know that for traitlets one has to also use the `@observe` decorator, not just `@default`, to implement the observer pattern, but that's about it.

Reply:
Should I use this? https://docs.python.org/dev/library/functools.html#functools.cached_property
Apparently it requires Python 3.8 though which in practice probably isn't a realistic dependency.

Reply:
OK I updated the commit to change the properties to (what should be according to StackOverflow) cached properties.